drivers/sensor/stm32_vbat/stm32_vbat.c

```diff
@@ -0,0 +1,126 @@
+/*
+ * Copyright (c) 2022 STMicroelectronics
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <device.h>
+#include <drivers/sensor.h>
+#include <drivers/adc.h>
+#include <logging/log.h>
+
+LOG_MODULE_REGISTER(stm32_vbat, CONFIG_SENSOR_LOG_LEVEL);
+
+#if DT_HAS_COMPAT_STATUS_OKAY(st_stm32_vbat)
+#define DT_DRV_COMPAT st_stm32_vbat
+#define HAS_CALIBRATION 0
+#else
+#error "No compatible devicetree node found"
+#endif
+
+struct stm32_vbat_data {
+	const struct device *adc;
+	uint8_t channel;
+	struct adc_channel_cfg adc_cfg;
+	struct adc_sequence adc_seq;
+	struct k_mutex mutex;
+	int16_t sample_buffer;
+	int16_t raw; /* raw adc Sensor value */
+};
+
+struct stm32_vbat_config {
+	int vbat_mv;
+	int ratio;
+};
+
+static int stm32_vbat_sample_fetch(const struct device *dev, enum sensor_channel chan)
+{
+	struct stm32_vbat_data *data = dev->data;
+	struct adc_sequence *sp = &data->adc_seq;
+	int rc;
+
+	if (chan != SENSOR_CHAN_ALL && chan != SENSOR_CHAN_VOLTAGE) {
+		return -ENOTSUP;
+	}
+
+	k_mutex_lock(&data->mutex, K_FOREVER);
+
+	rc = adc_read(data->adc, sp);
+	sp->calibrate = false;
+	if (rc == 0) {
+		data->raw = data->sample_buffer;
+	}
+
+	k_mutex_unlock(&data->mutex);
+
+	return 0;
+}
+
+static int stm32_vbat_channel_get(const struct device *dev, enum sensor_channel chan,
+				  struct sensor_value *val)
+{
+	struct stm32_vbat_data *data = dev->data;
+	const struct stm32_vbat_config *cfg = dev->config;
+	float voltage;
+
+	if (chan != SENSOR_CHAN_VOLTAGE) {
+		return -ENOTSUP;
+	}
+
+	voltage = data->raw * cfg->vbat_mv / 0x0FFF; /* Sensor value in millivolts */
+	/* considering the vbat input through a resistor bridge */
+	voltage = voltage * cfg->ratio;
+
+	return sensor_value_from_double(val, voltage);
+}
+
+static const struct sensor_driver_api stm32_vbat_driver_api = {
+	.sample_fetch = stm32_vbat_sample_fetch,
+	.channel_get = stm32_vbat_channel_get,
+};
+
+static int stm32_vbat_init(const struct device *dev)
+{
+	struct stm32_vbat_data *data = dev->data;
+	struct adc_channel_cfg *accp = &data->adc_cfg;
+	struct adc_sequence *asp = &data->adc_seq;
+	int rc;
+
+	k_mutex_init(&data->mutex);
+
+	if (!device_is_ready(data->adc)) {
+		LOG_ERR("Device %s is not ready", data->adc->name);
+		return -ENODEV;
+	}
+
+	*accp = (struct adc_channel_cfg){ .gain = ADC_GAIN_1,
+					  .reference = ADC_REF_INTERNAL,
+					  .acquisition_time = ADC_ACQ_TIME_MAX,
+					  .channel_id = data->channel,
+					  .differential = 0 };
+	rc = adc_channel_setup(data->adc, accp);
+	LOG_DBG("Setup AIN%u got %d", data->channel, rc);
+
+	*asp = (struct adc_sequence){
+		.channels = BIT(data->channel),
+		.buffer = &data->sample_buffer,
+		.buffer_size = sizeof(data->sample_buffer),
+		.resolution = 12,
+		.calibrate = true,
```

Comment:
Is the `calibrate` required? My testing shows that this makes no difference to the final result.

Reply:
no, removing